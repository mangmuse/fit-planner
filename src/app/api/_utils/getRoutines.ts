import { prisma } from "@/lib/prisma";

export const getRoutines = async (userId: string) => {
  const routines = await prisma.routine.findMany({ where: { userId } });
  return routines;
};

export const getRoutineIds = async (userId: string): Promise<string[]> => {
  const routines = await prisma.routine.findMany({
    where: { userId },
    select: { id: true },
  });
  const routineIds = routines.map((r) => r.id);
  return routineIds;
};
