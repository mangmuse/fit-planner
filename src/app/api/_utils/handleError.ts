import { NextResponse } from "next/server";

export class HttpError extends Error {
  public status: number;

  constructor(message: string, statusCode: number) {
    super(message);
    this.name = "HttpError";
    this.status = statusCode;

    Object.setPrototypeOf(this, HttpError.prototype);
  }
}

export function handleServerError(e: unknown) {
  console.error("[Server Error]", e);
  if (e instanceof HttpError) {
    return NextResponse.json(
      { success: false, message: e.message },
      { status: e.status }
    );
  }
  return NextResponse.json(
    { success: false, message: "예상치 못한 서버 에러" },
    { status: 500 }
  );
}
