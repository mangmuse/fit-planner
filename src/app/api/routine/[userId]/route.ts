import { getRoutines } from "@/app/api/_utils/getRoutines";
import { handleServerError } from "@/app/api/_utils/handleError";
import { validateData } from "@/util/validateData";
import { NextRequest, NextResponse } from "next/server";
import { z } from "zod";

export const GET = async (
  req: NextRequest,
  { params }: { params: Promise<{ userId: string }> }
) => {
  try {
    const { userId } = await Promise.resolve(params);
    const parsedUserId = validateData<string>(z.string(), userId);
    const routines = await getRoutines(parsedUserId);

    return NextResponse.json({ success: true, routines }, { status: 200 });
  } catch (e) {
    return handleServerError(e);
  }
};
