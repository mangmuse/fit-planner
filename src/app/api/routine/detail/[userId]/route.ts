import { getRoutineIds } from "@/app/api/_utils/getRoutines";
import { handleServerError } from "@/app/api/_utils/handleError";
import { prisma } from "@/lib/prisma";
import { validateData } from "@/util/validateData";
import { RoutineDetail } from "@prisma/client";
import { NextRequest, NextResponse } from "next/server";
import { z } from "zod";

export const GET = async (
  req: NextRequest,
  { params }: { params: Promise<{ userId: string }> }
) => {
  try {
    const { userId } = await Promise.resolve(params);
    const parsedUserId = validateData<string>(z.string(), userId);
    const routineIds = await getRoutineIds(parsedUserId);
    const routineDetails = await prisma.routineDetail.findMany({
      where: {
        routineId: {
          in: routineIds,
        },
      },
      include: {
        exercise: {
          select: {
            name: true,
          },
        },
      },
    });
    const transformedDetails: RoutineDetail[] = routineDetails.map((detail) => {
      const { exercise, ...rest } = detail;
      return { ...rest, exerciseName: exercise.name };
    });
    return NextResponse.json({
      success: true,
      routineDetails: transformedDetails,
    });
  } catch (e) {
    return handleServerError(e);
  }
};
