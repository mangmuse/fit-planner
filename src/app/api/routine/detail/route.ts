import { getRoutineIds } from "@/app/api/_utils/getRoutines";
import { handleServerError } from "@/app/api/_utils/handleError";
import { prisma } from "@/lib/prisma";
import { validateData } from "@/util/validateData";
import { Prisma, RoutineDetail } from "@prisma/client";
import { NextRequest, NextResponse } from "next/server";
import { z } from "zod";

export type RoutineDetailWithIncludes = Prisma.RoutineDetailGetPayload<{
  include: {
    exercise: {
      select: {
        name: true;
      };
    };
  };
}>;

export const GET = async (req: NextRequest) => {
  try {
    const userId = req.nextUrl.searchParams.get("userId");
    if (!userId || typeof userId !== "string") {
      return NextResponse.json(
        {
          success: false,
          message: "userId가 없거나 타입이 올바르지 않습니다",
        },
        { status: 400 }
      );
    }

    const parsedUserId = validateData<string>(z.string(), userId);
    const routineIds = await getRoutineIds(parsedUserId);
    const routineDetails = await prisma.routineDetail.findMany({
      where: {
        routineId: {
          in: routineIds,
        },
      },
      include: {
        exercise: {
          select: {
            name: true,
          },
        },
      },
    });
    const transformedDetails: RoutineDetail[] = routineDetails.map((detail) => {
      const { exercise, ...rest } = detail;
      return { ...rest, exerciseName: exercise.name };
    });
    return NextResponse.json({
      success: true,
      routineDetails: transformedDetails,
    });
  } catch (e) {
    return handleServerError(e);
  }
};
