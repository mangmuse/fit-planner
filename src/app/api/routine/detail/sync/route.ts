import pMap from "p-map";
import { SyncRoutineDetailsToServerResponse } from "@/api/routineDetail.api";
import { handleServerError, HttpError } from "@/app/api/_utils/handleError";
import {
  localRoutineDetailSchema,
  LocalRoutineDetailWithServerRoutineId,
} from "@/types/models";
import { validateData } from "@/util/validateData";
import { NextRequest, NextResponse } from "next/server";
import { z } from "zod";
import { prisma } from "@/lib/prisma";

const requestBodySchema = z.object({
  mappedUnsynced: z.array(
    localRoutineDetailSchema.extend({
      routineId: z.string(),
      id: z.number(),
    })
  ),
});

type RequestBody = z.infer<typeof requestBodySchema>;

const createRoutineDetail = async (
  detail: LocalRoutineDetailWithServerRoutineId
) => {
  const routine = await prisma.routine.findUnique({
    where: { id: detail.routineId },
  });
  if (!routine)
    throw new HttpError("RoutineId가 일치하는 routine을 찾지 못했습니다", 404);

  const created = await prisma.routineDetail.create({
    data: {
      weight: detail.weight,
      rpe: detail.rpe,
      reps: detail.reps,
      setOrder: detail.setOrder,
      setType: detail.setType,
      exerciseOrder: detail.exerciseOrder,
      routine: { connect: { id: detail.routineId } },
      exercise: { connect: { id: detail.exerciseId } },
    },
  });
  return created;
};

const updateRoutineDetail = async (
  detail: LocalRoutineDetailWithServerRoutineId,
  serverId: string
) => {
  const updated = await prisma.routineDetail.update({
    where: { id: serverId },
    data: {
      weight: detail.weight,
      rpe: detail.rpe,
      reps: detail.reps,
      setOrder: detail.setOrder,
      setType: detail.setType,
      exerciseOrder: detail.exerciseOrder,
    },
  });
  return updated;
};

export const POST = async (req: NextRequest) => {
  try {
    const body = await req.json();
    const parsedBody = validateData<RequestBody>(requestBodySchema, body);
    const unsynced = parsedBody.mappedUnsynced;

    const updatedList: SyncRoutineDetailsToServerResponse["updated"] =
      await pMap(
        unsynced,
        async (detail) => {
          let serverDetailId = detail.serverId;
          if (serverDetailId) {
            const updated = await updateRoutineDetail(detail, serverDetailId);
            serverDetailId = updated.id;
          } else {
            const created = await createRoutineDetail(detail);
            serverDetailId = created.id;
          }
          return {
            localId: detail.id,
            serverId: serverDetailId,
            exerciseId: detail.exerciseId,
            routineId: detail.routineId,
          };
        },
        { concurrency: 5 }
      );

    return NextResponse.json({ success: true, updated: updatedList });
  } catch (e) {
    return handleServerError(e);
  }
};
