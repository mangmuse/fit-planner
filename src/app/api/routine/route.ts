import { getRoutines } from "@/app/api/_utils/getRoutines";
import { handleServerError } from "@/app/api/_utils/handleError";
import { validateData } from "@/util/validateData";
import { NextRequest, NextResponse } from "next/server";
import { z } from "zod";

export const GET = async (req: NextRequest) => {
  try {
    const userId = req.nextUrl.searchParams.get("userId");
    if (!userId || typeof userId !== "string") {
      return NextResponse.json(
        {
          success: false,
          message: "userId가 없거나 타입이 올바르지 않습니다",
        },
        { status: 400 }
      );
    }
    const parsedUserId = validateData<string>(z.string(), userId);
    const routines = await getRoutines(parsedUserId);

    return NextResponse.json({ success: true, routines }, { status: 200 });
  } catch (e) {
    return handleServerError(e);
  }
};
