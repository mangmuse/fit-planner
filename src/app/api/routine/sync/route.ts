import { create } from "zustand";
import pMap from "p-map";
import { localRoutineSchema } from "@/types/models";
import { validateData } from "@/util/validateData";
import { NextRequest, NextResponse } from "next/server";
import { z } from "zod";
import { prisma } from "@/lib/prisma";
import { handleServerError } from "@/app/api/_utils/handleError";

const requestBodySchema = z.object({
  unsynced: z.array(localRoutineSchema),
});

type RequestBody = z.infer<typeof requestBodySchema>;

export const POST = async (req: NextRequest) => {
  try {
    const body = await req.json();
    const parsedBody = validateData<RequestBody>(requestBodySchema, body);
    const unsynced = parsedBody.unsynced;
    console.log(unsynced, "unsyncedunsyncedunsynced");
    const updatedList: Array<{ localId: number; serverId: string }> = [];
    await pMap(
      unsynced,
      async (routine) => {
        let routineRecord;
        if (routine.serverId) {
          const updatedRecord = await prisma.routine.update({
            where: {
              id: routine.serverId,
            },
            data: {
              name: routine.name,
              description: routine.description || "",
              updatedAt: routine.updatedAt,
            },
          });
          if (routine.id) {
            updatedList.push({
              localId: routine.id,
              serverId: updatedRecord.id,
            });
          }
        } else {
          const createdRecord = await prisma.routine.create({
            data: {
              userId: routine.userId,
              name: routine.name,
              description: routine.description || "",
              createdAt: routine.createdAt,
            },
          });
          if (routine.id) {
            updatedList.push({
              localId: routine.id,
              serverId: createdRecord.id,
            });
          }
        }
      },
      { concurrency: 5 }
    );
    return NextResponse.json({
      success: true,
      updated: updatedList,
    });
  } catch (e) {
    return handleServerError(e);
  }
};
