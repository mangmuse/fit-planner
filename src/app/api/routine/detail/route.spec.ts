jest.mock("@/app/api/_utils/getRoutines");
jest.mock("@/app/api/_utils/handleError");
jest.mock("@/util/validateData");
import { createPrismaRoutineDetailResponse } from "@/__mocks__/app/api/routine/detail/mockData";
import { getRoutineIds } from "@/app/api/_utils/getRoutines";
import { handleServerError } from "@/app/api/_utils/handleError";
import { GET, RoutineDetailWithIncludes } from "@/app/api/routine/detail/route";
import { prisma } from "@/lib/prisma";
import { ApiErrorResponse, ApiSuccessResponse } from "@/types/apiResponse";
import { ClientRoutineDetail } from "@/types/models";
import { validateData } from "@/util/validateData";
import { NextRequest, NextResponse } from "next/server";

describe("GET /api/routine/detail", () => {
  const mockedGetRoutineIds = getRoutineIds as jest.Mock;
  const mockedValidateData = validateData as jest.Mock;
  const mockedHandleServerError = handleServerError as jest.Mock;
  const mockedPrismaFindMany = prisma.routineDetail.findMany as jest.Mock;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  const targetUrl = "http://localhost/api/routine/detail";
  const mockUserId = "user-123";
  const getMockFindManyResponse = (): RoutineDetailWithIncludes[] => [
    createPrismaRoutineDetailResponse({ routineId: "routine-1" }),
    createPrismaRoutineDetailResponse({ routineId: "routine-2" }),
    createPrismaRoutineDetailResponse({ routineId: "routine-3" }),
  ];

  it("routineDetail이 없는경우 200 상태코드와 함께 빈 배열을 반환한다", async () => {
    const req = new NextRequest(`${targetUrl}?userId=${mockUserId}`);

    mockedValidateData.mockReturnValue(mockUserId);
    mockedGetRoutineIds.mockResolvedValue([]);
    mockedPrismaFindMany.mockResolvedValue([]);

    const res = await GET(req);
    const body = (await res.json()) as ApiSuccessResponse<{
      routineDetails: ClientRoutineDetail[];
    }>;

    expect(res.status).toBe(200);
    expect(body.routineDetails).toEqual([]);
  });

  it("userId가 없거나 올바르지 않은 타입이면 400 상태코드와 함께 에러 메시지를 반환한다", async () => {
    const req = new NextRequest(`${targetUrl}`);

    const res = await GET(req);
    const body = (await res.json()) as ApiErrorResponse;

    expect(res.status).toBe(400);
    expect(body.message).toBe("userId가 없거나 타입이 올바르지 않습니다");

    expect(mockedGetRoutineIds).not.toHaveBeenCalled();
    expect(mockedHandleServerError).not.toHaveBeenCalled();
  });

  it("DB 조회 중 오류가 발생하면 500 상태코드와 함께 에러 메시지를 반환한다", async () => {
    const req = new NextRequest(`${targetUrl}?userId=${mockUserId}`);
    const mockError = new Error("DB 조회 중 오류");

    mockedValidateData.mockReturnValue(mockUserId);
    mockedGetRoutineIds.mockResolvedValue([
      "routine-1",
      "routine-2",
      "routine-3",
    ]);

    mockedPrismaFindMany.mockRejectedValue(mockError);
    mockedHandleServerError.mockResolvedValue(
      NextResponse.json(
        { success: false, message: mockError.message },
        { status: 500 }
      )
    );

    const res = await GET(req);
    const body = (await res.json()) as ApiErrorResponse;

    expect(res.status).toBe(500);
    expect(body.message).toBe(mockError.message);
  });

  it("routineDetail 목록을 200 상태코드와 함께 반환한다", async () => {
    const req = new NextRequest(`${targetUrl}?userId=${mockUserId}`);

    mockedValidateData.mockReturnValue(mockUserId);
    mockedGetRoutineIds.mockResolvedValue([
      "routine-1",
      "routine-2",
      "routine-3",
    ]);
    mockedPrismaFindMany.mockResolvedValue(getMockFindManyResponse());

    const res = await GET(req);
    const body = (await res.json()) as ApiSuccessResponse<{
      routineDetails: ClientRoutineDetail[];
    }>;

    expect(res.status).toBe(200);
    expect(body.routineDetails).toMatchSnapshot();

    expect(mockedGetRoutineIds).toHaveBeenCalledWith(mockUserId);
  });
});
