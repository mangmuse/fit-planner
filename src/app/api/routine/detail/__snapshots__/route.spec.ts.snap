// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /api/routine/detail routineDetail 목록을 200 상태코드와 함께 반환한다 1`] = `
[
  {
    "createdAt": "2024-01-01T10:00:00.000Z",
    "exerciseId": 123,
    "exerciseName": "벤치프레스",
    "exerciseOrder": 1,
    "id": "routineDetail-1",
    "reps": 10,
    "routineId": "routine-1",
    "rpe": 10,
    "setOrder": 1,
    "setType": "NORMAL",
    "updatedAt": "2024-01-02T10:00:00.000Z",
    "weight": 100,
  },
  {
    "createdAt": "2024-01-01T10:00:00.000Z",
    "exerciseId": 123,
    "exerciseName": "벤치프레스",
    "exerciseOrder": 1,
    "id": "routineDetail-1",
    "reps": 10,
    "routineId": "routine-2",
    "rpe": 10,
    "setOrder": 1,
    "setType": "NORMAL",
    "updatedAt": "2024-01-02T10:00:00.000Z",
    "weight": 100,
  },
  {
    "createdAt": "2024-01-01T10:00:00.000Z",
    "exerciseId": 123,
    "exerciseName": "벤치프레스",
    "exerciseOrder": 1,
    "id": "routineDetail-1",
    "reps": 10,
    "routineId": "routine-3",
    "rpe": 10,
    "setOrder": 1,
    "setType": "NORMAL",
    "updatedAt": "2024-01-02T10:00:00.000Z",
    "weight": 100,
  },
]
`;
